src/cascadia/WindowsTerminal/WindowThread.cpp

```diff
@@ -52,15 +78,102 @@ void WindowThread::RundownForExit()
+//   they want to re-use this window thread for a new window.
+// Return Value:
+// - true IFF we should enter this thread's message loop
+// INVARIANT: This must be called on our "ui thread", our window thread.
+bool WindowThread::KeepWarm()
+{
+    if (_host != nullptr)
     {
-        MSG msg = {};
-        while (PeekMessageW(&msg, nullptr, 0, 0, PM_REMOVE))
+        // We're currently hot
+        return true;
+    }
+
+    // If we're refrigerated, then wait on the microwave signal, which will be
+    // raised when we get re-heated by another thread to reactivate us.
+
+    if (_warmWindow != nullptr)
+    {
+        std::unique_lock lock(_microwave);
+        _microwaveBuzzer.wait(lock);
+
+        // If ThrowAway() was called, then the buzzer will be signalled without
+        // setting a new _host. In that case, the app is quitting, for real. We
+        // just want to exit with false.
+        const bool reheated = _host != nullptr;
+        if (reheated)
         {
-            ::DispatchMessageW(&msg);
+            _UpdateSettingsRequestedToken = _host->UpdateSettingsRequested([this]() { _UpdateSettingsRequestedHandlers(); });
+            // Re-initialize the host here, on the window thread
+            _host->Initialize();
         }
+        return reheated;
+    }
+    else
+    {
+        return false;
     }
 }
 
+// Method Description:
+// - "Refrigerate" this thread for later reuse. This will refrigerate the window
+//   itself, and tear down our current app host. We'll save our window for
+//   later. We'll also pump out the existing message from XAML, before
+//   returning. After we return, the emperor will add us to the list of threads
+//   that can be re-used.
+void WindowThread::Refrigerate()
+{
+    _host->UpdateSettingsRequested(_UpdateSettingsRequestedToken);
+
+    // keep a reference to the HWND and DesktopWindowXamlSource alive.
+    _warmWindow = std::move(_host->Refrigerate());
+
+    // rundown remaining messages before destructing the app host
+    _pumpRemainingXamlMessages();
+    _host = nullptr;
+}
+
+// Method Description:
+// - "Reheat" this thread for reuse. We'll build a new AppHost, and pass in the
+//   existing window to it. We'll then trigger the _microwaveBuzzer, so KeepWarm
+//   (which is on the UI thread) will get unblocked, and we can initialize this
+//   window.
+void WindowThread::Microwave(
+    winrt::Microsoft::Terminal::Remoting::WindowRequestedArgs args,
+    winrt::Microsoft::Terminal::Remoting::Peasant peasant)
+{
+    _peasant = std::move(peasant);
+    _args = std::move(args);
+
+    _host = std::make_unique<::AppHost>(_appLogic,
+                                        _args,
+                                        _manager,
+                                        _peasant,
+                                        std::move(_warmWindow));
+
+    // raise the signal to unblock KeepWarm and start the window message loop again.
+    _microwaveBuzzer.notify_one();

```

Comment:
I just remember that C++20 added `std::binary_semaphore` and we could use it in place of a conditional variable. Instead of `notify_one()` it'd be [`release()`](https://en.cppreference.com/w/cpp/thread/counting_semaphore/release) and instead of `wait()` it'd be [`acquire()`](https://en.cppreference.com/w/cpp/thread/counting_semaphore/acquire).